File: Java_Data_Structures_And_Algorithms/Sorting_Algorithms/Quicksort.java
class QuickSort
{
    int partition(int a[], int low, int high)
    {
        int pivot = a[high]; 
        int i = (low-1);
        for (int j=low; j<high; j++)
        {
            if (a[j] <= pivot)
            {
                i++;

                int temp = arr[i];
                a[i] = a[j];
                a[j] = temp;
            }
        }

        int temp = a[i+1];
        a[i+1] = a[high];
        a[high] = temp;

        return i+1;
    }
    void sort(int a[], int l, int h)
    {
        if (l < h)
        {
            int pi = partition(a, l, h);
            sort(a, l, pi-1);
            sort(a, pi+1, h);
        }
    }
    public static void main(String args[])
    {
        int a[] = {1, 7, 2, 4, 6, 5};
        int n = a.length;

        QuickSort ob = new QuickSort();
        ob.sort(a, 0, n-1);

        for (int i=0; i<n; ++i)
            System.out.print(a[i]+" ");
    }
}

File: C/Sorting/QuickSort.c

#include <stdio.h>
int Partition(int A[],int low,int high){
    int pivot=A[low],left=low,right=high;
    while(left<right){
        while(left<high && A[left]<=pivot){
            left++;
        }
        while(right>low && A[right]>pivot){
            right--;
        }
        if(left<right){
            int temp = A[left];
            A[left]=A[right];
            A[right]=temp;
        }
    
    }
    int temp2=A[low];
    A[low]=A[right];
    A[right]=temp2;
    return right;
}

void quickSort(int A[],int low,int high){
    if(low<high){
        int p= Partition(A,low,high);
        quickSort(A,low,p-1);
        quickSort(A,p+1,high);
    }
}


int main(){
    int n,a[100],i,p;
    printf("\nEnter the number of elements:");
    scanf("%d",&n);

    //Input the array elements
    printf("Enter the array elements:");
    for ( i=0 ; i<n ; i++){
        scanf("%d",a+i);
    }
    // Show elements in a array before sorting
    printf("Elements in the array before sorting:");
    for( i=0 ; i<n ; i++){
        printf("%d ",a[i]);
    }
    printf("\n");
    quickSort(a,0,n-1);
    // Show elements in a array after sorting
    printf("Elements in the array after  quick sort:");
    for( i=0 ; i<n ; i++){
        printf("%d ",a[i]);
    }
    printf("\n");
}
// Time Complexity:O(N^2) Average case:Theta(logN)